core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java

```diff
@@ -93,16 +93,41 @@ public synchronized  InodeFile getInodeFile() throws FileDoesNotExistException {
    */
   public synchronized  InodeDirectory getParentInodeDirectory()
       throws InvalidPathException, FileDoesNotExistException {
-    if (mPathComponents.length < 2 || mInodes.size() < (mPathComponents.length - 1)) {
-      // The path is only the root, or the list of inodes is not long enough to contain the parent
+    InodeDirectory inodeDirectory = (InodeDirectory) peekInode(false);
+    if (inodeDirectory == null) {
       throw new FileDoesNotExistException(
           ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(mUri.getParent()));
     }
-    Inode<?> inode = mInodes.get(mPathComponents.length - 2);
-    if (!inode.isDirectory()) {
-      throw new InvalidPathException(ExceptionMessage.PATH_MUST_HAVE_VALID_PARENT.getMessage(mUri));
+    return inodeDirectory;
+  }
+
+  /**
+   * Peeks an inode on the inode path. The choice of which inode depends on whether the request
+   * comes from a command with recursive flag or not.
+   *
+   * @param isRecursive true if this peekInode is invoked by a command with recursive flag
+   * @return the last existing inode on the inode path if isRecursive is true, the second
+   *         last inode on the inode path if isRecursive is false
+   * @throws InvalidPathException if the parent exists but is not a directory
+   */
+  public synchronized Inode peekInode(boolean isRecursive)
```

Comment:
This method is not very intuitive to me. Peek usually means it returns the last (or first) item from a data structure, without removing that item. What does "peek" mean w.r.t. a locked path? To me, it sounds like "peek" should just return the last inode, if it exists.

Also, `isRecursive` doesn't makes sense to me either. How is recursive related to peeking an inode in a path? I feel like recursive doesn't mean anything w.r.t. an existing inode path. Recursive is usually a concept related to creating sequence of new inodes.